File: android/src/main/java/com/simplesoundplayer/SimpleSoundPlayerTurboModule.java
package com.simplesoundplayer;

import androidx.annotation.NonNull;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.module.annotations.ReactModule;
import android.media.MediaPlayer;

@ReactModule(name = SimpleSoundPlayerModule.NAME)
public class SimpleSoundPlayerModule extends ReactContextBaseJavaModule {
    public static final String NAME = "SimpleSoundPlayer";
    private MediaPlayer mediaPlayer;
    private final ReactApplicationContext reactContext;

    public SimpleSoundPlayerModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @NonNull
    @Override
    public String getName() {
        return NAME;
    }

    // Método para reproducir un sonido
    @ReactMethod
    public void play(String soundName) {
        if (mediaPlayer != null) {
            mediaPlayer.release();
        }

        // Obtiene el recurso de audio desde "raw" en los recursos de Android
        int resId = reactContext.getResources().getIdentifier(soundName, "raw", reactContext.getPackageName());
        if (resId != 0) {
            mediaPlayer = MediaPlayer.create(reactContext, resId);
            mediaPlayer.setOnCompletionListener(mp -> destroy());
            mediaPlayer.setOnErrorListener((mp, what, extra) -> {
                destroy();
                return true;
            });
            mediaPlayer.start();
        }
    }

    // Método para detener la reproducción actual
    @ReactMethod
    public void stop() {
        if (mediaPlayer != null && mediaPlayer.isPlaying()) {
            mediaPlayer.stop();
            mediaPlayer.reset();
        }
    }

    // Método para destruir el reproductor y liberar recursos
    @ReactMethod
    public void destroy() {
        if (mediaPlayer != null) {
            if (mediaPlayer.isPlaying()) {
                mediaPlayer.stop();
            }
            mediaPlayer.release();
            mediaPlayer = null;
        }
    }

    // Método para ajustar el volumen de reproducción
    @ReactMethod
    public void setVolume(float volume) {
        if (mediaPlayer != null) {
            mediaPlayer.setVolume(volume, volume);
        }
    }
}

File: ios/SimpleSoundPlayerTurboModule.m
// Archivo: SimpleSoundPlayerTurboModule.m

#import "SimpleSoundPlayerTurboModule.h"
#import <React/RCTUtils.h>

@implementation SimpleSoundPlayerTurboModule

RCT_EXPORT_MODULE(SimpleSoundPlayerTurboModule)

// Método para reproducir un sonido
- (void)play:(NSString *)soundName resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject {
    NSError *error = nil;

    // Obtener la ruta del archivo de sonido
    NSString *filePath = [[NSBundle mainBundle] pathForResource:soundName ofType:nil];
    if (!filePath) {
        reject(@"file_not_found", [NSString stringWithFormat:@"No se encontró el archivo de sonido: %@", soundName], nil);
        return;
    }

    NSURL *soundURL = [NSURL fileURLWithPath:filePath];
    self.audioPlayer = [[AVAudioPlayer alloc] initWithContentsOfURL:soundURL error:&error];

    if (error) {
        reject(@"playback_error", @"Error al reproducir el sonido.", error);
        return;
    }

    self.audioPlayer.delegate = self;
    [self.audioPlayer prepareToPlay];
    if ([self.audioPlayer play]) {
        resolve(@(YES));
    } else {
        reject(@"playback_failed", @"No se pudo iniciar la reproducción.", nil);
    }
}

// Método para detener la reproducción
- (void)stop {
    if (self.audioPlayer.isPlaying) {
        [self.audioPlayer stop];
    }
}

// Método para ajustar el volumen
- (void)setVolume:(nonnull NSNumber *)volume {
    if (self.audioPlayer) {
        self.audioPlayer.volume = [volume floatValue];
    }
}

// Método para destruir y liberar recursos del reproductor de audio
- (void)destroy {
    if (self.audioPlayer) {
        [self.audioPlayer stop];
        self.audioPlayer = nil;
    }
}

// Delegado para la finalización de la reproducción
- (void)audioPlayerDidFinishPlaying:(AVAudioPlayer *)player successfully:(BOOL)flag {
    [self sendEventWithName:@"onPlaybackEnded" body:@{@"success": @(flag)}];
}

// Manejo de errores de decodificación de audio
- (void)audioPlayerDecodeErrorDidOccur:(AVAudioPlayer *)player error:(NSError *)error {
    [self sendEventWithName:@"onPlaybackError" body:@{@"error": error.localizedDescription}];
}

// Método para enviar eventos a JavaScript
- (void)sendEventWithName:(NSString *)name body:(NSDictionary *)body {
    if (self.bridge) {
        [self sendEventWithName:name body:body];
    }
}

@end
